server/src/Utils/Order.hs

```diff
@@ -0,0 +1,38 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Utils.Order where
+
+import Import.NoFoundation
+import Model.Types
+
+-- {-| Determine if a given user is allowed to access the PDF file of an item.
+-- -}
+hasAccessToPdfFileDownload ::
+       ( BaseBackend (YesodPersistBackend site) ~ SqlBackend
+       , YesodPersist site
+       , PersistQueryRead (YesodPersistBackend site)
+       )
+    => UserId
+    -> Text
+    -> HandlerT site IO AuthResult
+hasAccessToPdfFileDownload userId filename = do
+    let unauthorized = Unauthorized "You didn't buy this item"
```

Comment:
I see. One way to do this would to incorporate `Maybe` into the monad you're using in this `do` block via a monad transformer. I was hoping I might not have to say 'monad transformer', but I think this would actually be the cleanest solution, since your `do` bindings would handle the `Maybe` automatically. You'd have to add `transformers` as a dependency and add a `lift` somewhere at the beginning of the function, but otherwise there wouldn't be much to it.

If you'd rather not think about monad transformers, there would probably be an alternative, but I'd have to think a little more … in any case, any alternative solution would be less seamless but more explicit—that's the tradeoff